File: src/tcp/messages/SetChannelUriMessage.py
from tcp.messages.AbstractMessage import AbstractMessage
from typing import Any

class SetChannelUriMessage(AbstractMessage):
    ACTION: str = "set_channel_uri"

    channelUri: str
    deviceId: str

    def __init__(self, jsonObj: Any):
        super(SetChannelUriMessage, self).__init__(jsonObj = jsonObj)

    def _fromJson(self, jsonObj: Any):
        super()._fromJson(jsonObj)
        self.channelUri = jsonObj["channel_uri"]
        self.deviceId = jsonObj["device_id"]

    def _toJson(self):
        jsonObj: Any = super()._toJson()
        jsonObj["channel_uri"] = self.channelUri
        jsonObj["device_id"] = self.deviceId
        return jsonObj

File: src/tcp/messages/SetChannelUriMessage.cpp
#include "SetChannelUriMessage.hpp"
#include "logger/Logger.hpp"

namespace tcp::messages {
SetChannelUriMessage::SetChannelUriMessage(const nlohmann::json& j) {
    // NOLINTNEXTLINE (clang-analyzer-optin.cplusplus.VirtualCall) Not relevant in this case
    isValid = from_json(j);
}

SetChannelUriMessage::SetChannelUriMessage(std::string&& channelUri, std::string&& deviceId) : AbstractMessage(std::string{ACTION}), channelUri(std::move(channelUri)), deviceId(std::move(deviceId)) {}

const std::string& SetChannelUriMessage::get_channel_uri() const { return channelUri; }

const std::string& SetChannelUriMessage::get_device_id() const { return deviceId; }

AbstractMessage::MessageType SetChannelUriMessage::get_type() const { return MessageType::SET_CHANNEL_URI_MESSAGE; }

bool SetChannelUriMessage::from_json(const nlohmann::json& j) {
    if (!AbstractMessage::from_json(j)) {
        return false;
    }

    if (!j.contains("channel_uri")) {
        LOG_WARNING << "Missing 'channel_uri' field in message.";
        return false;
    }
    channelUri = j["channel_uri"];
    if (channelUri.empty()) {
        LOG_WARNING << "Invalid message 'channel_uri' value. Expected a non empty string, but received: " << channelUri;
        return false;
    }

    if (!j.contains("device_id")) {
        LOG_WARNING << "Missing 'device_id' field in message.";
        return false;
    }
    deviceId = j["device_id"];
    if (deviceId.empty()) {
        LOG_WARNING << "Invalid message 'device_id' value. Expected a non empty string, but received: " << deviceId;
        return false;
    }
    return true;
}

void SetChannelUriMessage::to_json(nlohmann::json& j) const {
    AbstractMessage::to_json(j);
    j["device_id"] = deviceId;
    j["channel_uri"] = channelUri;
}

}  // namespace tcp::messages